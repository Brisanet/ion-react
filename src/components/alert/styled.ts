import stitches from '../../stitches.config';
import { getFillBgFontBorderColors } from '../utils';
import { spacing } from '../utils/spacing';

const { styled } = stitches;

export const AlertStyled = styled('div', {
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  padding: `${spacing(1)} ${spacing(2)} ${spacing(1)} ${spacing(1.5)}`,
  gap: spacing(1),

  minHeight: 24,

  backgroundColor: '$positive1',

  borderWidth: '1px 1px 1px 8px',
  borderStyle: 'solid',
  borderColor: '$positive6',
  borderRadius: 8,

  fontSize: 14,
  fontWeight: 400,
  color: '$neutral8',

  svg: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: spacing(1),

    fill: '$primary6',

    borderRadius: 9,
  },

  span: {
    width: '95%',
  },

  '& svg:last-child': {
    fill: '$primary6',
  },

  variants: {
    type: {
      info: {
        ...getFillBgFontBorderColors({
          backgroundColor: '$info1',
          borderColor: '$info6',
          svgFillColor: '$info6',
          fontColor: '$neutral8',
        }),
      },
      warning: {
        ...getFillBgFontBorderColors({
          backgroundColor: '$warning1',
          borderColor: '$warning6',
          svgFillColor: '$warning6',
          fontColor: '$neutral8',
        }),
      },
      negative: {
        ...getFillBgFontBorderColors({
          backgroundColor: '$negative1',
          borderColor: '$negative6',
          svgFillColor: '$negative6',
          fontColor: '$neutral8',
        }),
      },
      success: {
        ...getFillBgFontBorderColors({
          backgroundColor: '$positive1',
          borderColor: '$positive6',
          svgFillColor: '$positive6',
          fontColor: '$neutral8',
        }),
      },
    },
    hideBackground: {
      true: {
        backgroundColor: 'transparent !important',
        borderColor: 'transparent !important',
      },
    },
  },
});
