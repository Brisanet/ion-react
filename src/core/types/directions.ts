export type BorderDirection = 'top' | 'right' | 'bottom' | 'left';
