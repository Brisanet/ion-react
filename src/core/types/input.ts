export type InputType = 'password' | 'text' | 'number' | 'email';
