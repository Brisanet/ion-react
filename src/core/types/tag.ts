import { iconType } from '../../components/icons/svgs/icons';
import { TagStatus } from './status';

export interface IonTagProps {
  outline?: boolean;
  status?: TagStatus;
  color?: string;
  label: string;
  icon?: iconType;
}
