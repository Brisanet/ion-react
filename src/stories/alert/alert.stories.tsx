import { ComponentStory, ComponentMeta } from '@storybook/react';

import IonAlert, { AlertProps } from '../../components/alert/alert';

export default {
  title: 'Ion/Feedback/Alert',
  component: IonAlert,
  argTypes: {
    handClick: { action: 'clicked' },
    type: {
      name: 'type',
      defaultValue: 'success',
      control: 'radio',
      options: ['success', 'info', 'warning', 'negative'],
    },
  },
} as ComponentMeta<typeof IonAlert>;

const Template: ComponentStory<typeof IonAlert> = (args: AlertProps) => (
  <IonAlert {...args} />
);

export const AlertDefault = Template.bind({});
AlertDefault.args = {
  message: 'Example of alert message',
  type: 'success',
  hideBackground: false,
  closable: false,
};
