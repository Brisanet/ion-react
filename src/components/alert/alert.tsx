import { StatusType } from '../../core/types/status';
import { IonButton } from '../button';
import ErrorBoundary from '../error/error-boundary';

import { IonIcon } from '../icons/icons';
import isValidLabel from '../utils/isValidLabel';
import { Alert, Wrapper } from './styled';

export interface AlertProps {
  message: string;
  type?: StatusType;
  closable?: boolean;
  hideBackground?: boolean;
  onClose?: () => void;
}

type iconType =
  | 'check-solid'
  | 'exclamation-solid'
  | 'info-solid'
  | 'close-solid';

const icons: Record<StatusType, iconType> = {
  info: 'info-solid',
  warning: 'exclamation-solid',
  negative: 'close-solid',
  success: 'check-solid',
};

const SIZE_ICON = 24;

const getIcon = (alertType: StatusType): iconType => icons[alertType];

export const IonAlert = ({
  message,
  type = 'success',
  closable = false,
  hideBackground = false,
  onClose,
}: AlertProps) => {
  if (!isValidLabel(message)) {
    return <ErrorBoundary msg='Message cannot be empty' />;
  }

  return (
    <Alert
      data-testid='ion-alert'
      $type={type}
      $hideBackground={hideBackground}
      $closable={closable}
    >
      <Wrapper>
        <IonIcon type={getIcon(type)} size={SIZE_ICON}></IonIcon>
        <span>{message}</span>
      </Wrapper>
      {closable && (
        <IonButton
          circular
          variant='ghost'
          icon='close'
          size='sm'
          onClick={onClose}
        />
      )}
    </Alert>
  );
};
