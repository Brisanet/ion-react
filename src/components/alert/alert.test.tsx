import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { IonAlert, AlertProps } from './alert';
import { StatusType } from '../../core/types/status';

const defaultAlert: AlertProps = {
  message: 'Example message',
};

const sut = (props = defaultAlert) => render(<IonAlert {...props} />);
const alertId = 'ion-alert';
const getAlert = () => screen.getByTestId(alertId);

describe('IonAlert', () => {
  it('should render alert', () => {
    sut();
    expect(getAlert()).toBeTruthy();
  });

  it('should render alert with label "Message alert"', async () => {
    const customLabel = 'Message alert';
    sut({ message: customLabel });
    expect(await screen.findAllByText(customLabel)).toBeTruthy();
  });

  it('should render ErrorBoundary component when not exist message', async () => {
    sut({ message: '' });
    const msgError = 'Message cannot be empty';
    const errorBoundary = screen.getByTestId('ion-error-boundary');
    expect(errorBoundary).toBeTruthy();
    expect(await screen.findByText(msgError)).toBeTruthy();
  });

  it.each(['success', 'info', 'warning', 'negative'] as StatusType[])(
    'should render alert with type: %s',
    (type) => {
      sut({ ...defaultAlert, type });
      expect(getAlert().className).toContain(`type-${type}`);
    }
  );

  it('should not render background alert when hideBackground is true', async () => {
    await sut({ ...defaultAlert, hideBackground: true });
    expect(getAlert().className).toContain('hideBackground-true');
  });

  it('should render icon close when closable is true', async () => {
    await sut({ ...defaultAlert, closable: true });
    const iconClose = screen.getByTestId('ion-icon-close');
    expect(iconClose).toBeTruthy();
  });

  it('should remove alert when click on close icon', async () => {
    await sut({ ...defaultAlert, closable: true });
    const iconClose = screen.getByTestId('ion-icon-close');
    expect(getAlert()).toBeTruthy();

    await userEvent.click(iconClose);
    expect(screen.queryByTestId(alertId)).not.toBeTruthy();
  });
});
