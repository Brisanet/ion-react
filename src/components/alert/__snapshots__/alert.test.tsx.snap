// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IonAlert should render alert with type: info 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #282b33;
  border-radius: 8px;
  height: 4rem;
  padding: 8px 16px 8px 12px;
  border: 1px solid #0bb2cb;
  background-color: #e2f9fd;
  border-left: 8px solid #0bb2cb;
}

.c0 .c1 svg {
  fill: #0bb2cb;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
}

<div>
  <div
    class="c0"
    data-testid="ion-alert"
  >
    <div
      class="c1 c2"
    >
      <svg
        data-testid="ion-icon-info-solid"
        fill="#282B33"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
         
        <path
          clip-rule="evenodd"
          d="M12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3ZM12 18C11.4477 18 11 17.5523 11 17V11C11 10.4477 11.4477 10 12 10C12.5523 10 13 10.4477 13 11V17C13 17.5523 12.5523 18 12 18ZM10.75 7.25C10.75 6.55964 11.3096 6 12 6C12.6904 6 13.25 6.55964 13.25 7.25C13.25 7.94036 12.6904 8.5 12 8.5C11.3096 8.5 10.75 7.94036 10.75 7.25Z"
          fill-rule="evenodd"
        />
         
      </svg>
      <span>
        Example message
      </span>
    </div>
  </div>
</div>
`;

exports[`IonAlert should render alert with type: negative 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #282b33;
  border-radius: 8px;
  height: 4rem;
  padding: 8px 16px 8px 12px;
  border: 1px solid #d6293a;
  background-color: #faeaec;
  border-left: 8px solid #d6293a;
}

.c0 .c1 svg {
  fill: #d6293a;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
}

<div>
  <div
    class="c0"
    data-testid="ion-alert"
  >
    <div
      class="c1 c2"
    >
      <svg
        data-testid="ion-icon-close-solid"
        fill="#282B33"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
         
        <path
          clip-rule="evenodd"
          d="M5.63604 18.364C9.15076 21.8787 14.8492 21.8787 18.364 18.364C21.8787 14.8492 21.8787 9.15076 18.364 5.63604C14.8492 2.12132 9.15076 2.12132 5.63604 5.63604C2.12132 9.15076 2.12132 14.8492 5.63604 18.364ZM15.5355 9.87868C15.9261 9.48816 15.9261 8.85499 15.5355 8.46447C15.145 8.07394 14.5118 8.07394 14.1213 8.46447L12 10.5858L9.87868 8.46447C9.48816 8.07394 8.85499 8.07394 8.46447 8.46447C8.07394 8.85499 8.07394 9.48816 8.46447 9.87868L10.5858 12L8.46447 14.1213C8.07394 14.5118 8.07394 15.145 8.46447 15.5355C8.85499 15.9261 9.48816 15.9261 9.87868 15.5355L12 13.4142L14.1213 15.5355C14.5118 15.9261 15.145 15.9261 15.5355 15.5355C15.9261 15.145 15.9261 14.5118 15.5355 14.1213L13.4142 12L15.5355 9.87868Z"
          fill-rule="evenodd"
        />
         
      </svg>
      <span>
        Example message
      </span>
    </div>
  </div>
</div>
`;

exports[`IonAlert should render alert with type: success 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #282b33;
  border-radius: 8px;
  height: 4rem;
  padding: 8px 16px 8px 12px;
  border: 1px solid #2d9f70;
  background-color: #e7f9f1;
  border-left: 8px solid #2d9f70;
}

.c0 .c1 svg {
  fill: #2d9f70;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
}

<div>
  <div
    class="c0"
    data-testid="ion-alert"
  >
    <div
      class="c1 c2"
    >
      <svg
        data-testid="ion-icon-check-solid"
        fill="#282B33"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
         
        <path
          clip-rule="evenodd"
          d="M12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21ZM16.7071 10.7071C17.0976 10.3166 17.0976 9.68342 16.7071 9.29289C16.3166 8.90237 15.6834 8.90237 15.2929 9.29289L11 13.5858L8.70711 11.2929C8.31658 10.9024 7.68342 10.9024 7.29289 11.2929C6.90237 11.6834 6.90237 12.3166 7.29289 12.7071L10.2929 15.7071C10.4804 15.8946 10.7348 16 11 16C11.2652 16 11.5196 15.8946 11.7071 15.7071L16.7071 10.7071Z"
          fill-rule="evenodd"
        />
         
      </svg>
      <span>
        Example message
      </span>
    </div>
  </div>
</div>
`;

exports[`IonAlert should render alert with type: warning 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #282b33;
  border-radius: 8px;
  height: 4rem;
  padding: 8px 16px 8px 12px;
  border: 1px solid #f9a915;
  background-color: #fff5e0;
  border-left: 8px solid #f9a915;
}

.c0 .c1 svg {
  fill: #f9a915;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
}

<div>
  <div
    class="c0"
    data-testid="ion-alert"
  >
    <div
      class="c1 c2"
    >
      <svg
        data-testid="ion-icon-attention-solid"
        fill="#282B33"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
         
        <path
          clip-rule="evenodd"
          d="M12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3C7.02944 3 3 7.02944 3 12C3 16.9706 7.02944 21 12 21ZM12 6C12.5523 6 13 6.44772 13 7V13C13 13.5523 12.5523 14 12 14C11.4477 14 11 13.5523 11 13V7C11 6.44772 11.4477 6 12 6ZM13.25 16.75C13.25 17.4404 12.6904 18 12 18C11.3096 18 10.75 17.4404 10.75 16.75C10.75 16.0596 11.3096 15.5 12 15.5C12.6904 15.5 13.25 16.0596 13.25 16.75Z"
          fill-rule="evenodd"
        />
         
      </svg>
      <span>
        Example message
      </span>
    </div>
  </div>
</div>
`;
